sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -263,6 +268,20 @@ public Mono<Utils.ValueHolder<AddressInformation[]>> tryGetAddresses(RxDocumentS
                         this.suboptimalServerPartitionTimestamps.putIfAbsent(partitionKeyRangeIdentity, Instant.now());
                     }
 
+                    // Refresh the cache if there was an address has been marked as unhealthy long enough and need to revalidate its status
+                    // If you are curious about why we do not depend on 410 to force refresh the addresses, the reason being:
+                    // When an address is marked as unhealthy, then the address enumerator will move it to the end of the list
+                    // So it could happen that no request will use the unhealthy address for an extended period of time
+                    // So the 410 -> forceRefresh workflow may not happen
+                    if (Arrays
+                            .stream(addressesValueHolder.v)
+                            .anyMatch(addressInformation -> addressInformation.getPhysicalUri().shouldRefreshHealthStatus())) {
```

Comment:
In `shouldRefreshHealthStatus()` function, we compare like this -> 

```
public boolean shouldRefreshHealthStatus() {
        return this.healthStatus.get() == HealthStatus.Unhealthy
                && Instant.now().compareTo(this.lastUnhealthyTimestamp.plusMillis(DEFAULT_NON_HEALTHY_RESET_TIME_IN_MILLISECONDS)) > 0;
    }
```
Shouldn't we compare to >= 0 (otherwise we might miss extra 60 seconds for refresh if the Instant.now() is equal to the last unhealthy time + refresh interval.

Reply:
make sense, will update in next iterator

Reply:
updated